### PR TITLE
CUDA: add FP32 FlashAttention vector kernel

### LINKED ISSUE #7055: Flash Attention not working with NVIDIA Quadro P3200 Pascal Architecture GPU
I am using the server of llama.cpp [b2781](https://github.com/ggerganov/llama.cpp/releases/tag/b2781) on a Windows machine with enabled `--flash-attn` option.

The Compute Capability of the Quadro P3200 GPU in the machine is 6.1. The installed NVIDIA Driver is at 552.22 and CUDA 12.3.

I am getting the following error message:

> ERROR: CUDA kernel flash_attn_ext_f16 has no device code compatible with CUDA arch 610. ggml-cuda.cu was compiled for: 520,610,700

**Log**

```
system_info":"AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 0 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | "}
llama_model_loader: loaded meta data with 24 key-value pairs and 291 tensors from .\vendor\llama.cpp\models\openchat-3.5-0106.Q5_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = D:\Arbeit\windows_manage_large_langua...
llama_model_loader: - kv   2:                           llama.vocab_size u32              = 32002
llama_model_loader: - kv   3:                       llama.context_length u32              = 8192
llama_model_loader: - kv   4:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   5:                          llama.block_count u32              = 32
llama_model_loader: - kv   6:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   7:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   8:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   9:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  10:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  11:                       llama.rope.freq_base f32              = 10000.000000
llama_model_loader: - kv  12:                          general.file_type u32              = 17
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  14:                      tokenizer.ggml.tokens arr[str,32002]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  15:                      tokenizer.ggml.scores arr[f32,32002]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,32002]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  17:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  18:                tokenizer.ggml.eos_token_id u32              = 32000
llama_model_loader: - kv  19:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  20:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  21:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  22:                    tokenizer.chat_template str              = {{ bos_token }}{% for message in mess...
llama_model_loader: - kv  23:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type q5_K:  193 tensors
llama_model_loader: - type q6_K:   33 tensors
llm_load_vocab: special tokens definition check successful ( 261/32002 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 32002
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: n_ctx_train      = 8192
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: n_embd_k_gqa     = 1024
llm_load_print_meta: n_embd_v_gqa     = 1024
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 10000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 8192
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 8B
llm_load_print_meta: model ftype      = Q5_K - Medium
llm_load_print_meta: model params     = 7.24 B
llm_load_print_meta: model size       = 4.78 GiB (5.67 BPW)
llm_load_print_meta: general.name     = D:\Arbeit\windows_manage_large_language_models\source
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 32000 '<|end_of_turn|>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: Quadro P3200 with Max-Q Design, compute capability 6.1, VMM: yes
llm_load_tensors: ggml ctx size =    0.30 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =    85.94 MiB
llm_load_tensors:      CUDA0 buffer size =  4807.06 MiB
...................................................................................................
llama_new_context_with_model: n_ctx      = 8192
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 1
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =  1024.00 MiB
llama_new_context_with_model: KV self size  = 1024.00 MiB, K (f16):  512.00 MiB, V (f16):  512.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.24 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =    88.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    24.01 MiB
llama_new_context_with_model: graph nodes  = 903
llama_new_context_with_model: graph splits = 2
[1714720299] warming up the model with an empty run
D:\Arbeit\windows_llama.cpp\vendor\llama.cpp\ggml-cuda\fattn.cu:571: ERROR: CUDA kernel flash_attn_ext_f16 has no device code compatible with CUDA arch 610. ggml-cuda.cu was compiled for: 520,610,700
[...]
D:\Arbeit\windows_llama.cpp\vendor\llama.cpp\ggml-cuda\fattn.cu:571: ERROR: CUDA kernel flash_attn_ext_f16 has no device code compatible with CUDA arch 610. ggml-cuda.cu was compiled for: 520,610,700
ggml_cuda_compute_forward: SILU failed
CUDA error: unspecified launch failure
  current device: 0, in function ggml_cuda_compute_forward at D:\Arbeit\windows_llama.cpp\vendor\llama.cpp\ggml-cuda.cu:2305
  err
GGML_ASSERT: D:\Arbeit\windows_llama.cpp\vendor\llama.cpp\ggml-cuda.cu:61: !"CUDA error"
```

Flash Attention is working for me on a Turing GPU.

**Question:** Should llama.cpp support Pascal GPUs? 

If not it would be nice to have a feature detection and a speaking error message that indicates this.

### PR DESCRIPTION
**Pull request by JohannesGaessler**
This PR adds an FP32 FlashAttention kernel that is very similar to the FP16 kernel. It enables using FlashAttention on NVIDIA GPUs without fast FP16 and without tensor cores. It should also provide a speedup on more recent NVIDIA GPUs for batch size 1 and FP32 precision. I have moved the FP16 and FP32 FlashAttention vector kernels to separate files in order to speed up compilation. I also added a function `ggml_backend_cuda_get_device_cc` to `ggml-cuda.h` in order to avoid breaking `tests/test-backend-ops` on NVIDIA GPUs without tensor cores. Unlike with the FP16 kernel there are no weird issues with arrays of size 1 vs. regular variables.

Performance on 1x P40:

| model                          | backend    | ngl |    n_batch |         fa | test       | t/s              |
| ------------------------------ | ---------- | --: | ---------: | ---------: | ---------- | ---------------: |
| llama 8B Q4_0                  | CUDA       |  99 |          1 |          0 | pp 4096    | 40.17 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          1 |          1 | pp 4096    | 52.69 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          2 |          0 | pp 4096    | 42.57 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          2 |          1 | pp 4096    | 97.77 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          4 |          0 | pp 4096    | 71.13 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          4 |          1 | pp 4096    | 117.34 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |          8 |          0 | pp 4096    | 83.96 ± 0.00     |
| llama 8B Q4_0                  | CUDA       |  99 |          8 |          1 | pp 4096    | 143.52 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         16 |          0 | pp 4096    | 110.77 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         16 |          1 | pp 4096    | 125.14 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         32 |          0 | pp 4096    | 210.72 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         32 |          1 | pp 4096    | 209.30 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         64 |          0 | pp 4096    | 387.33 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |         64 |          1 | pp 4096    | 315.04 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        128 |          0 | pp 4096    | 532.72 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        128 |          1 | pp 4096    | 357.28 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        256 |          0 | pp 4096    | 664.68 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        256 |          1 | pp 4096    | 374.81 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        512 |          0 | pp 4096    | 748.74 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |        512 |          1 | pp 4096    | 375.69 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       1024 |          0 | pp 4096    | 749.28 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       1024 |          1 | pp 4096    | 375.78 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       2048 |          0 | pp 4096    | 749.41 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       2048 |          1 | pp 4096    | 375.89 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       4096 |          0 | pp 4096    | 749.52 ± 0.00    |
| llama 8B Q4_0                  | CUDA       |  99 |       4096 |          1 | pp 4096    | 375.90 ± 0.00    |

### CODE REVIEW

**Comment by JohannesGaessler** [comment]
Fixes https://github.com/ggerganov/llama.cpp/issues/7055 .

**Comment by slaren** [comment]
This happens regularly, but it's never going to be ok to add backend-specific functions to `test-backend-ops`. Instead, add the necessary checks to the `supports_op` function in ggml-cuda.

**Comment by sorasoras** [comment]
I wonder how would this work on a rdna3 GPU compare to rdna2.

**Comment by sorasoras** [comment]
```
  Device 0: AMD Radeon RX 7900 XTX, compute capability 11.0, VMM: no
| model                          |       size |     params | backend    | ngl | sm         |         fa | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------: | ---------- | ---------------: |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | pp 512     |   687.77 ± 12.71 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | tg 128     |     34.70 ± 0.30 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | pp 512     |    767.92 ± 1.57 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | tg 128     |     34.36 ± 0.14 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | ROCm       |  99 | none       |          1 | pp 512     |   1511.97 ± 9.18 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | ROCm       |  99 | none       |          1 | tg 128     |     57.05 ± 0.02 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | ROCm       |  99 | none       |          0 | pp 512     |   1773.69 ± 5.63 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | ROCm       |  99 | none       |          0 | tg 128     |     56.31 ± 0.72 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | pp 1024    |    650.39 ± 8.36 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | pp 2048    |    574.70 ± 3.05 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | pp 4096    |    465.18 ± 3.77 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          1 | tg 128     |     35.06 ± 0.06 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | pp 1024    |   760.63 ± 11.81 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | pp 2048    |    726.50 ± 7.10 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | pp 4096    |    669.02 ± 2.67 |
| qwen2 ?B IQ4_XS - 4.25 bpw     |  16.51 GiB |    32.51 B | ROCm       |  99 | none       |          0 | tg 128     |     33.90 ± 0.30 |
build: de85f908 (2834)

```

```
  Device 0: Tesla P40, compute capability 6.1, VMM: yes
| model                          |       size |     params | backend    | ngl | sm         |         fa | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------: | ---------- | ---------------: |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 |          1 | pp 512     |    180.33 ± 0.31 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 |          1 | tg 128     |     11.35 ± 0.01 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 |          0 | pp 512     |    201.03 ± 0.31 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 |          0 | tg 128     |      9.31 ± 0.02 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       |          1 | pp 512     |    376.66 ± 0.05 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       |          1 | tg 128     |     22.66 ± 0.03 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       |          0 | pp 512     |    435.80 ± 0.16 |
| qwen2 13B Q5_K - Small         |   9.33 GiB |    14.17 B | CUDA       |  99 | none       |          0 | tg 128     |     17.88 ± 0.02 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          1 | pp 1024    |    166.17 ± 0.07 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          1 | pp 2048    |    143.93 ± 0.05 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          1 | pp 4096    |    113.58 ± 0.09 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          1 | tg 128     |     11.30 ± 0.00 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          0 | pp 1024    |    196.44 ± 0.25 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          0 | pp 2048    |    189.21 ± 0.23 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          0 | pp 4096    |    177.17 ± 0.21 |
| qwen2 ?B Q4_K - Medium         |  18.34 GiB |    32.51 B | CUDA       |  99 | none       |          0 | tg 128     |      9.31 ± 0.01 |
build: de85f908 (2834)

```

TG speed  up is significant but PP is slower quite a bit, I don't know why.

**Comment by JohannesGaessler** [comment]
There simply isn't yet a kernel optimized for large batch sizes.

File: ggml-cuda.cu [hunk 1]
```diff
@@ -2841,8 +2841,21 @@ GGML_CALL static bool ggml_backend_cuda_supports_op(ggml_backend_t backend, cons
...
+            return op->src[0]->ne[0] == 64 || op->src[0]->ne[0] == 128;
+#else
+            if (op->src[0]->ne[0] == 64 || op->src[0]->ne[0] == 128) {
+                return true;
+            }
+            for (int id = 0; id < ggml_backend_cuda_get_device_count(); ++id) {
+                if (ggml_cuda_info().devices[id].cc < CC_VOLTA) {
+                    return false;
+                }
+            }
```
Comment (slaren) [inline ggml-cuda.cu:2856]: I don't think it is necessary to check every device here, instead get the context and check only the device for this context. Something like this:
```C++
ggml_backend_cuda_context * cuda_ctx = (ggml_backend_cuda_context *)backend->context;
if (ggml_cuda_info().devices[ctx->device].cc < CC_VOLTA) {
    return false;
}
```

**Comment by scottmudge** [comment]
Just adding a small data point, with KoboldCPP compiled with this, with a Q8_K 11b model on 2 x 1080 Ti (Pascal) setup, I get:

* ~20.2 T/s avg (proc + gen) with FP32 FA enabled.
* ~13.4 T/s avg (proc + gen) with FP32 FA disabled.

So a significant improvement in my case. Whereas with FP16 FA, I saw a decrease. So it definitely has utility for a subset of users.

**Comment by JohannesGaessler** [comment]
I don't have any ALiBi models set up for testing but according to `tests/test-backend-ops` the implementation works correctly.

**Review by slaren — APPROVED** [review]

**Comment by JohannesGaessler** [comment]
Pascal is still supported, make an issue.

**Pull request merged**